Support handle invalid logs

baurine opened the pull request:
close #17 , close #16 

What did:

- Support handle the log files that include invalid logs in the beginning and end, try at most 10 lines to get the first valid log.
- Support handle the log files that include invalid logs in the middle, treat the invalid log with previous log time, level, and message by the whole line content.
- Fix the bug that would read incorrect content from the end of the log file when others append new contents to the file
- Improve the performance for reading content from the end of the log file, read 512 bytes once instead of 1 byte

baurine commented:
@breeswish @crazycs520 @lonng PTAL, thanks!

baurine commented on search_log.go:
```diff
@@ -136,20 +175,22 @@ func readLine(reader *bufio.Reader) (string, error) {
 
-func readLastLine(file *os.File) string {
+// Read a line from the end of a file.
+// TODO: read byte by byte is low efficiency, we can improve it, for example, read 1024 bytes one time
```
Can we refine by this way? (not sure whether it has some pitfalls.)

crazycs520 replied:
Sure, we can add a todo here.

baurine commented on search_log.go:
```diff
@@ -136,20 +175,22 @@ func readLine(reader *bufio.Reader) (string, error) {
 	for {
-		cursor -= 1
+		cursor--
 		file.Seek(cursor, io.SeekEnd)
```
What if the log has appended some new logs after the file is opened when calling `file.seek(0, io.SeekEnd)`.

lonng replied:
Nice catch, I think we should add some tests for these scenarios.

baurine replied:
ok, I will do some tests to check it.

baurine replied:
Hi @lonng , add a test method `TestReadAndAppendLogFile` to test this case, it seems we will read the unexcepted result from the end when new contents is appending to the file. But we can try to resolve it in another PR.

![企业微信20200707120345](https://user-images.githubusercontent.com/1284531/86706958-06a7e200-c04a-11ea-8740-e55fde82adf4.png)

baurine replied:
will continue to fix it in this PR.

baurine replied:
Fixed it by `file.Seek(cursor, io.SeekStart)`.

breeswish commented:
> Support handle the log files that include invalid logs in the beginning and end, skip them until find a valid log.

Maybe we can set a threshold for this behaviour, i.e. maximum look ahead xxx bytes, in case of the whole big file is invalid logs and causing performance degrade

> Support handle the log files that include invalid logs in the middle, treat the invalid log with previous log time, unknown log level, and message by the whole line content.

Maybe we can keep the same log time, log level as the previous one

baurine commented:
> > Support handle the log files that include invalid logs in the beginning and end, skip them until find a valid log.
> 
> Maybe we can set a threshold for this behaviour, i.e. maximum look ahead xxx bytes, in case of the whole big file is invalid logs and causing performance degrade

There is a `tryLines` parameter to control the maximum lines to read, but currently `tryLines` is set to 0 to mean may read all lines, and no way to config it from outside.

> > Support handle the log files that include invalid logs in the middle, treat the invalid log with previous log time, unknown log level, and message by the whole line content.
> 
> Maybe we can keep the same log time, log level as the previous one

ok

baurine commented on search_log.go:
```diff
@@ -136,34 +175,40 @@ func readLine(reader *bufio.Reader) (string, error) {
+	var cursor = endCursor
 	stat, _ := file.Stat()
 	filesize := stat.Size()
+	fmt.Println("current filesize:", filesize)
```
Will remove it later

baurine commented on search_log_test.go:
```diff
@@ -515,3 +534,102 @@ func (s *searchLogSuite) TestParseLogItem(c *C) {
+	fmt.Println("final content:\n", string(content))
+}
+
+// result: (may be different every time)
```
Will remove later

crazycs520 commented on search_log_test.go:
```diff
@@ -515,3 +534,102 @@ func (s *searchLogSuite) TestParseLogItem(c *C) {
+
+	stat, _ := file.Stat()
+	filesize := stat.Size()
+	fmt.Println("initial filesize: ", filesize)
```
remove this?

crazycs520 commented on search_log_test.go:
```diff
@@ -515,3 +534,102 @@ func (s *searchLogSuite) TestParseLogItem(c *C) {
+			line := fmt.Sprintf("[2020/07/07 06:%d:17.011 -04:00] [INFO] appended logs\n", i)
+			_, err := file_append.WriteString(line)
+			c.Assert(err, IsNil, Commentf("write %s failed", line))
+			fmt.Print("write line: ", line)
```
ditto?

crazycs520 commented on search_log_test.go:
```diff
@@ -515,3 +534,102 @@ func (s *searchLogSuite) TestParseLogItem(c *C) {
 	}
 }
+
+func (s *searchLogSuite) TestReadAndAppendLogFile(c *C) {
```
This test output is not expected. I think we need to re-implement the `readLineReverse` function.

baurine replied:
Yep, I decided to continue to re-implement it in this PR.

baurine replied:
Done!

baurine commented on search_log.go:
```diff
@@ -136,34 +176,37 @@ func readLine(reader *bufio.Reader) (string, error) {
+		}
 
+		cursor--
+		file.Seek(cursor, io.SeekStart)
```
Replace `file.Seek(cursor, io.SeekEnd)` by `file.Seek(cursor, io.SeekStart)`, the former way may read incorrect content from the end of a file when some others append new logs to this file.

baurine commented:
Addressed all comments, polished by the following work:

- Fix the bug that would read incorrect content from the end of the log file when others append new contents to the file
- Improve the performance for reading content from the end of the log file, read 512 bytes once instead of 1 byte

@breeswish @lonng @crazycs520 PTAL, thanks!

baurine commented on search_log.go:
```diff
@@ -136,34 +177,45 @@ func readLine(reader *bufio.Reader) (string, error) {
-		if cursor == -filesize { // stop if we are at the begining
-			break
+
+		var size int64 = 512
```
Read 512 bytes once instead of 1 byte

breeswish replied:
I like this change very much, would you like to write a benchmark for this optimization?

baurine replied:
Comparison:

```
Old way:
searchLogSuite.BenchmarkReadLastLine        10000            124423 ns/op
searchLogSuite.BenchmarkReadLastLine        10000            126135 ns/op

New way:
searchLogSuite.BenchmarkReadLastLines      1000000              1892 ns/op
searchLogSuite.BenchmarkReadLastLines      1000000              1920 ns/op
```

Conclusion: the new implementation is 65x (126135/1920) as fast as the old way.

The benchmark for the readLastLine in PR #20

breeswish commented on search_log.go:
```diff
@@ -109,18 +103,65 @@ func resolveFiles(logFilePath string, beginTime, endTime int64) ([]logFile, erro
+			break
+		}
+		endCursor -= int64(len(lines))
+		linesSlice := strings.FieldsFunc(lines, func(c rune) bool { return c == '\n' || c == '\r' })
```
Maybe there will be problems when line ending is `\r\n`.

baurine replied:
Yep, I thought that, let me try to polish it.

baurine replied:
Polish it by `strings.Split(strings.ReplaceAll(finalStr, "\r\n", "\n"), "\n")` and put it inside the `readLastLines` method.

breeswish commented on search_log.go:
```diff
@@ -136,34 +177,45 @@ func readLine(reader *bufio.Reader) (string, error) {
-			break
+
+		var size int64 = 512
+		if cursor < size {
+			size = cursor
+		}
+		cursor -= size
```
I would prefer not changing the size, but changing the cursor instead:

```suggestion
		if cursor < size {
			cursor = 0
		} else {
			cursor -= size
		}
```

breeswish commented on search_log.go:
```diff
@@ -136,34 +177,45 @@ func readLine(reader *bufio.Reader) (string, error) {
+		cursor -= size
+
+		file.Seek(cursor, io.SeekStart)
+		chars := make([]byte, size)
```
We can move this line out of the loop and reuse the buffer to improve performance

baurine replied:
Good idea! update it soon.

baurine replied:
I tried but found it doesn't work well, including keep size value not change. Because `file.Read(chars)` reads `len(chars)` bytes content, so we need to make sure the chars is allocated by the size we need to read, instead of fixed 512 bytes.

breeswish commented on search_log.go:
```diff
@@ -136,34 +177,45 @@ func readLine(reader *bufio.Reader) (string, error) {
+		lines = append(chars, lines...)
+
+		// find first '\n' or '\r'
+		for i := 0; i < len(chars); i++ {
```
We only need to check `\r` and `\n` in the scanned bytes from this round, instead of previous rounds

baurine replied:
Yes, it works as you said, `len(chars)` is the length of content just read in this round.

breeswish commented:
Very cool!

crazycs520 commented on search_log.go:
```diff
@@ -136,34 +177,45 @@ func readLine(reader *bufio.Reader) (string, error) {
-	filesize := stat.Size()
+// Read lines from the end of a file
+// endCursor initial value should be the filesize
+func readLastLines(file *os.File, endCursor int64) string {
```
`readLastLines` look like maybe returns multiple lines, but the return parameter is `string`, should be `[]string`?

baurine replied:
I need to get read bytes by `len(string)` in its caller method, return `[]string` will miss the length info, else I need to add an extra `readSize` return value. Which way is better?

crazycs520 replied:
I prefer `lines []string, size int`, it's more clear

baurine replied:
ok!

baurine replied:
Done!

breeswish commented on search_log.go:
```diff
@@ -274,21 +332,33 @@ nextLine:
+				continue
+			}
+			// handle invalid log
+			// make whole line as log message with pre time and unknown log_level
```
```suggestion
			// make whole line as log message with pre time and pre log_level
```
